File: android/src/androidTest/java/org/ligi/gobandroidhd/base/GoViewActions.java
package org.ligi.gobandroidhd.base;

import android.support.test.espresso.UiController;
import android.support.test.espresso.matcher.ViewMatchers;
import android.view.View;

import android.support.test.espresso.ViewAction;
import android.support.test.espresso.action.CoordinatesProvider;
import android.support.test.espresso.action.GeneralClickAction;
import android.support.test.espresso.action.Press;
import android.support.test.espresso.action.Tap;
import android.widget.SeekBar;

import org.hamcrest.Matcher;
import org.ligi.gobandroid_hd.App;
import org.ligi.gobandroid_hd.logic.Cell;

public class GoViewActions {

    public static ViewAction placeStone(final Cell cell) {
        return new GeneralClickAction(
                Tap.SINGLE,
                new CoordinatesProvider() {
                    @Override
                    public float[] calculateCoordinates(View view) {

                        final int[] screenPos = new int[2];
                        view.getLocationOnScreen(screenPos);
                        final int gameSize = App.getGame().getSize();

                        final float screenX = screenPos[0] + (0.5f + cell.x) * (view.getWidth() / gameSize);
                        final float screenY = screenPos[1] + (0.5f + cell.y) * (view.getHeight() / gameSize);

                        return new float[]{screenX, screenY};
                    }
                },
                Press.FINGER);
    }

    public static ViewAction tapStone(Cell cell) {
        return placeStone(cell);
    }

    public static ViewAction setProgress(final int progress) {
        return new ViewAction() {
            @Override
            public void perform(UiController uiController, View view) {
                SeekBar seekBar = (SeekBar) view;
                seekBar.setProgress(progress);
            }
            @Override
            public String getDescription() {
                return "Set a progress on a SeekBar";
            }
            @Override
            public Matcher<View> getConstraints() {
                return ViewMatchers.isAssignableFrom(SeekBar.class);
            }
        };
    }
}

File: android/src/androidTest/java/org/ligi/gobandroid_hd/test_helper_functions/GoViewActions.kt
package org.ligi.gobandroid_hd.test_helper_functions

import androidx.test.espresso.UiController
import androidx.test.espresso.ViewAction
import androidx.test.espresso.action.CoordinatesProvider
import androidx.test.espresso.action.GeneralClickAction
import androidx.test.espresso.action.Press
import androidx.test.espresso.action.Tap
import androidx.test.espresso.matcher.ViewMatchers
import android.view.View
import android.widget.SeekBar

import org.hamcrest.Matcher
import org.ligi.gobandroid_hd.logic.Cell
import org.ligi.gobandroid_hd.logic.GoGame


fun placeStone(cell: Cell, game: GoGame): ViewAction {
    return GeneralClickAction(
            Tap.SINGLE,
            CoordinatesProvider { view ->
                val screenPos = IntArray(2)
                view.getLocationOnScreen(screenPos)
                val gameSize = game.size

                val screenX = screenPos[0] + (0.5f + cell.x) * (view.width / gameSize)
                val screenY = screenPos[1] + (0.5f + cell.y) * (view.height / gameSize)

                floatArrayOf(screenX, screenY)
            },
            Press.FINGER)
}

fun tapStone(cell: Cell, game: GoGame): ViewAction {
    return placeStone(cell, game)
}

fun setProgress(progress: Int): ViewAction {
    return object : ViewAction {
        override fun perform(uiController: UiController, view: View) {
            val seekBar = view as SeekBar
            seekBar.progress = progress
        }

        override fun getDescription(): String {
            return "Set a progress on a SeekBar"
        }

        override fun getConstraints(): Matcher<View> {
            return ViewMatchers.isAssignableFrom(SeekBar::class.java)
        }
    }
}
